ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java

```diff
@@ -0,0 +1,596 @@
+  private final Random random = new Random();
+  private final FutureResult<ReadableStreamChannel> operationFuture = new FutureResult<>();
+
+  private int requestParallelism = 2;
+  private int successTarget = 1;
+  private int maxChunkSize;
+  private int blobSize = 1000;
+  private RouterConfig routerConfig;
+  private MockClusterMap mockClusterMap;
+  private NonBlockingRouterMetrics routerMetrics;
+  private MockServerLayout mockServerLayout;
+  private int replicasCount;
+  private AtomicReference<MockSelectorState> mockSelectorState = new AtomicReference<>();
+  private ResponseHandler responseHandler;
+  private MockNetworkClientFactory networkClientFactory;
+  private NetworkClient networkClient;
+  private VerifiableProperties vprops;
+  private NonBlockingRouter router;
+  private String blobIdStr;
+  private BlobProperties blobProperties;
+  private byte[] userMetadata;
+  private byte[] putContent;
+
+  private final GetTestRequestRegistrationCallbackImpl requestRegistrationCallback =
+      new GetTestRequestRegistrationCallbackImpl();
+
+  private class GetTestRequestRegistrationCallbackImpl implements RequestRegistrationCallback<GetOperation> {
+    private List<RequestInfo> requestListToFill;
+
+    @Override
+    public void registerRequestToSend(GetOperation getOperation, RequestInfo requestInfo) {
+      requestListToFill.add(requestInfo);
+      correlationIdToGetOperation.put(((RequestOrResponse) requestInfo.getRequest()).getCorrelationId(), getOperation);
+    }
+  }
+
+  private final AtomicInteger operationsCount = new AtomicInteger(0);
+  private final OperationCompleteCallback operationCompleteCallback = new OperationCompleteCallback(operationsCount);
+
+  public GetBlobOperationTest() {
+    // Defaults. Tests should override these as appropriate.
+    maxChunkSize = random.nextInt(1024 * 1024) + 1;
+    blobSize = maxChunkSize * random.nextInt(10) + random.nextInt(maxChunkSize - 1) + 1;
+  }
+
+  @After
+  public void after() {
+    if (networkClient != null) {
+      networkClient.close();
+    }
+    Assert.assertEquals("All operations should have completed", 0, operationsCount.get());
+  }
+
+  /**
+   * Instantiate a router, perform a put, close the router. The blob that was put will be saved in the MockServer,
+   * and can be queried by the getBlob operations in the test.
+   */
+  private void doPut()
+      throws Exception {
+    requestParallelism = 3;
+    successTarget = 2;
+    mockSelectorState.set(MockSelectorState.Good);
+    VerifiableProperties vprops = new VerifiableProperties(getNonBlockingRouterProperties());
+    routerConfig = new RouterConfig(vprops);
+    mockClusterMap = new MockClusterMap();
+    routerMetrics = new NonBlockingRouterMetrics(mockClusterMap.getMetricRegistry());
+    mockServerLayout = new MockServerLayout(mockClusterMap);
+    replicasCount = mockClusterMap.getWritablePartitionIds().get(0).getReplicaIds().size();
```

Comment:
The assumption is that all partitions have the same number of replicas. That is a fair assumption I think, as things like `successTarget` and `requestParallelism` are not partition specific.
That way, I think it belongs in the class and does not have to be computed on the fly.